Add token pass-thru for AuthConfig

dhiltgen opened the pull request:
This change allows API clients to retrieve an authentication token from
a registry, and then pass that token directly to the API.

Example usage:

```
REPO_USER=dhiltgen
read -s PASSWORD
REPO=privateorg/repo
AUTH_URL=https://auth.docker.io/token
TOKEN=$(curl -s -u "${REPO_USER}:${PASSWORD}" "${AUTH_URL}?scope=repository:${REPO}:pull&service=registry.docker.io" |
    jq -r ".token")

HEADER=$(echo "{\"registrytoken\":\"${TOKEN}\"}"|base64 -w 0 )
curl -s -D - -H "X-Registry-Auth: ${HEADER}" -X POST "http://localhost:2376/images/create?fromImage=${REPO}"
```

Signed-off-by: Daniel Hiltgen daniel.hiltgen@docker.com

Commit:
Add token pass-thru for AuthConfig

This change allows API clients to retrieve an authentication token from
a registry, and then pass that token directly to the API.

Example usage:

    REPO_USER=dhiltgen
    read -s PASSWORD
    REPO=privateorg/repo
    AUTH_URL=https://auth.docker.io/token
    TOKEN=$(curl -s -u "${REPO_USER}:${PASSWORD}" "${AUTH_URL}?scope=repository:${REPO}:pull&service=registry.docker.io" |
        jq -r ".token")

    HEADER=$(echo "{\"registrytoken\":\"${TOKEN}\"}"|base64 -w 0 )
    curl -s -D - -H "X-Registry-Auth: ${HEADER}" -X POST "http://localhost:2376/images/create?fromImage=${REPO}"

Signed-off-by: Daniel Hiltgen <daniel.hiltgen@docker.com>

calavera commented:
@docker/security what do you think about this?

NathanMcCauley commented:
This makes a lot of sense to me. Design LGTM.

calavera commented:
`Token` sounds like a very generic name to me. Should we use something more specific, like `RegistryToken` or `AuthToken`? I'm open to suggestions, and to keep it like this if other people agree.

NathanMcCauley commented:
I like `RegistryToken` in order to distinguish it from `AuthToken` which will be confused with #15365 and #17412.

dhiltgen commented:
Name change sounds fine to me.  If nobody else speaks up, I'll update this and the related PRs to make the change.

LK4D4 commented:
LGTM

calavera commented:
Is there a way to test this?

PS: We definitely need documentation before merging it :smile_cat:

dhiltgen commented:
Unit test and API level documentation added.  I wasn't sure if there was someplace better to doc it, since AuthConfig wasn't really described in any of the docs I found.

dhiltgen commented:
Not sure what to make of the CI build failures - janky seems to be a GPG key download failure.  The windows one is a bit harder to make heads or tails of, but I don't see anything related to my change in the log.

dhiltgen commented:
Hmm... looks like the graph package just went poof - so I need to figure out where this logic needs to shift to.

aaronlehmann commented:
This code moved to the `distribution` package.

thaJeztah commented:
@dhiltgen yup, huge PR was just merged for content addressability; https://github.com/docker/docker/pull/17924, apologies for that :relaxed:

dhiltgen commented:
Thanks all - I've migrated my change over to the new layout, and am just running some final tests now to make sure nothing broke in the transition - I'll update the PR shortly.

dhiltgen commented:
Updated based on new layout.

calavera commented:
LGTM

thaJeztah commented:
not in docs review yet, but can you also add a note to the API changelog; https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v122-api-changes

tonistiigi commented on distribution/registry.go:
```diff
@@ -94,7 +95,12 @@ func NewV2Repository(repoInfo *registry.RepositoryInfo, endpoint registry.APIEnd
 	creds := dumbCredentialStore{auth: authConfig}
```
Can we avoid creating these objects if they are not needed now?

Or even may be more readable if this just sets up array of handlers and there is only a single call to `NewAuthorizer`.

dhiltgen replied:
Yeah, I can shift these down into the else block to clean it up a bit.

dhiltgen commented:
Updated to clean up the handler code a little, and added the new field to the version change doc.

The userns CI failure seems unrelated to my change.

tonistiigi commented:
LGTM

abronan commented:
LGTM (but needs a rebase, sorry @dhiltgen)

thaJeztah commented on docs/reference/api/docker_remote_api_v1.22.md:
```diff
@@ -1529,7 +1529,20 @@ Query Parameters:
+
+    {
+        "registrytoken": "9cbaf023786cd7..."
+    }
```
the code examples probably need to be indented more (4 spaces) to show up as code blocks, see the rendered page;

<img width="693" alt="screen shot 2015-12-02 at 20 09 12" src="https://cloud.githubusercontent.com/assets/1804568/11541110/bf104634-9930-11e5-8cc8-2796cb01fd76.png">

thaJeztah commented on docs/reference/api/docker_remote_api_v1.22.md:
```diff
@@ -1738,8 +1751,20 @@ Query Parameters:
+
+    {
+        "registrytoken": "9cbaf023786cd7..."
+    }
```
Same here

dhiltgen commented:
Rebased and got markdown to cooperate for the code blocks.  (additional indentation didn't seem to work so I resorted to the triple-back-tick)

thaJeztah commented:
Thanks, @dhiltgen, markdown can be a PITA with whitespace and lists. I ran `make docs` on your PR and it rendered correctly now \o/

docs LGTM

ping @jamtur01 @SvenDowideit @moxiegirl for review

moxiegirl commented:
LGTM

thaJeztah commented:
Thanks @dhiltgen!

vdemeester commented:
:tada:

pdevine commented:
:fireworks:

stevvooe commented:
A little disappointing that this wasn't reviewed by anyone from the distribution team. 

There are few issues:

1. Tokens are typically treated as short-lived. They can be long-lived, but most of the fallback flow doesn't assume this.
2. How will this handle authentication failures that require the resolution of further permissions? There are protocols in V2 (using 401 status codes) that can instruct the client to request further permissions.

thaJeztah commented:
@stevvooe please don't comment on closed tickets :trollface: 

In your opinion, do we need changes to this PR before 1.10? If so, I can open an issue for that. Apologies that distribution didn't fully review this

stevvooe commented:
> @stevvooe please don't comment on closed tickets :trollface:

This is one of the few cases where it's acceptable. :hamster: 

> In your opinion, do we need changes to this PR before 1.10? If so, I can open an issue for that. Apologies that distribution didn't fully review this

I have no clue. We have no proposal showing why this is needed, other solutions considered, how this will interoperate with the existing push/pull flow or how we can support this going into the future.

I'd like to have seen a more expressive approach that includes exchanging a token on garant for a registry token. This would also have solved the unencrypted credentials issue. The difference being we can issue tokens that have require a higher level of protection and thus can have a longer expiry.

stevvooe commented on distribution/registry.go:
```diff
@@ -113,3 +119,16 @@ func digestFromManifest(m *schema1.SignedManifest, localName string) (digest.Dig
+}
+
+func (th *existingTokenHandler) AuthorizeRequest(req *http.Request, params map[string]string) error {
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", th.token))
```
After some thought, we may need to restrict the token to a particular host. Otherwise, we risk credential leak upon redirect. Also, tokens should also be included _only_ on https requests, unless otherwise specified.
